README.txt
= TwoWaySQL

=== sites:
* {Web Site}[http://twowaysql.rubyforge.org/]
* {Project Page}[http://rubyforge.org/projects/twowaysql]
* {API Doc}[http://twowaysql.rubyforge.org/rdoc/]
* {Issue Tracking}[http://twowaysql.rubyforge.org/issues/]
* {Coverage Report}[http://twowaysql.rubyforge.org/coverage/]


=== sources:
* http://github.com/twada/twowaysql/tree/master



== DESCRIPTION:


=== What is TwoWaySQL

'TwoWaySQL' is a concept, looks like a Template Engine for SQL.

It is initially invented and implemented in Seasar project's S2Dao[http://s2dao.seasar.org/en/index.html].

This package is a Ruby implementation of TwoWaySQL concept.


=== Why TwoWaySQL

Like any other technology, SQL is also in 80:20 world. 80% of SQL can be generated easily and automatically by O-R Mappers, 20% can not. 20% of your SQLs may large and complex, using CASE clause, self JOIN, UNION ALL, EXCEPT, ... these are the real strength of SQL and its set-based operations.

We better use them, to get the most out of RDBMS power.

TwoWaySQL encourages to write complex SQL manually. Just develop SQL using tools like pgAdmin3 in try and error style, then markup the SQL by TwoWaySQL. Usage and Features are in this doc.



=== Advantage
TwoWaySQL provides better separation of host language and SQL.

With TwoWaySQL, you can
* separate SQL (as file) from host language
* bind variables to SQL using Interpolation comments
* modify SQL conditionally by using Directive comments
* run and preview TwoWaySQL-style SQL by tools like pgAdmin3, since the SQL is still valid SQL.



== One minute example

  # given SQL string with TwoWaySQL comments
  sql = <<-EOS
    SELECT * FROM emp
    /*BEGIN*/WHERE
      /*IF ctx[:job]*/ job = /*ctx[:job]*/'CLERK' /*END*/
      /*IF ctx[:deptno_list]*/ AND deptno IN /*ctx[:deptno_list]*/(20, 30) /*END*/
      /*IF ctx[:age]*/ AND age > /*ctx[:age]*/30 /*END*/
    /*END*/
    /*IF ctx[:order_by] */ ORDER BY /*$ctx[:order_by]*/id /*$ctx[:order]*/ASC /*END*/
  EOS


  # parse the SQL to create template object
  template = TwoWaySQL::Template.parse(sql)


  # merge data with template
  data = {
    :age => 35,
    :deptno_list => [10,20,30],
    :order_by => 'age',
    :order => 'DESC'
  }
  merged = template.merge(data)


  expected_sql = <<-EOS
    SELECT * FROM emp
     WHERE
      
      deptno IN (?, ?, ?)
      AND age > ?

     ORDER BY age DESC
  EOS

  merged.sql == expected_sql      #=> true
  merged.bound_variables          #=> [10,20,30,35]


  # use merged SQL and variables with any O-R Mapper you like (ex. Sequel)
  require 'sequel'
  DB = Sequel.connect('postgres://user:pass@localhost:5432/mydb')
  rows = DB.fetch(merged.sql, *merged.bound_variables).all
  . . .


==== What TwoWaySQL intended to do
* TwoWaySQL is intended to be a small and simple module.
* TwoWaySQL respects SQL and its set-based operations. TwoWaySQL assists writing complex SQL with ease.
* TwoWaySQL is not a replacement of ActiveRecord,Sequel,or any other O-R Mappers. Instead, TwoWaySQL will work with O-R Mappers well as a SQL construction module.


==== What TwoWaySQL is not
TwoWaySQL is not
* an O-R Mapper
* a SQL Parser
* a framework
* a Prepared Statement



== FEATURES/PROBLEMS:

* Interpolation comments
  * Bind variable comment
  * Embedded variable comment

* Directive comments
  * IF comment
  * ELSE comment
  * BEGIN comment

* actual SQL comment

* parse options
  * preserve_space
  * preserve_comment



== SYNOPSIS:

NOTE: some of this section is based on docs for S2Dao[http://s2dao.seasar.org/en/s2dao.html#SQLBind]


=== Published Classes

TwoWaySQL::Template is the class you may only use. TwoWaySQL::Template acts as a Facade for this package, others are for internal use.


=== Basic Usage

==== Input
* TwoWaySQL-style SQL(string,file or anything like IO) to TwoWaySQL::Template.parse to create template object (note: template object is stateless and reentrant, so you can cache it)
  * (Optionally) TwoWaySQL::Template.parse accepts Hash of parse options as second argument
* data object to the TwoWaySQL::Template#merge then TwoWaySQL will evaluate the data as name 'ctx'.

==== Output
* SQL String with placeholders (generally, '?' is used for placeholders)
* Array of bound variables for placeholders



=== SQL comment

Firstly, In TwoWaySQL, expressions are written within SQL comment such as within "/**/" and "--". SQL may still be executed since TwoWaySQL specific, non-SQL expressions are written within comments. As a best practice, it is better to first write and test SQL and then write expressions within comments.

To write actual comments in SQL, *insert a space* after "/*" before the comment string. For example, /* foo*/. TwoWaySQL will recognize the space(s) after the comment start ("/*") and treat the enclosed content as an actual comment.



=== Bind variable comment

Bind variable comment is used to bind value(s) to the SQL.
Literal to the right of bind variable comment is automatically replaced with a value.
Bind variable comment syntax is as follows:

  /*variable_name*/Literal

TwoWaySQL may use bind variable as follows. In this case, value of ctx[:empno] is automatically set. Data object that is passed to TwoWaySQL::Template#merge is evaled as name 'ctx'.

  SELECT * FROM emp WHERE empno = /*ctx[:empno]*/7788

===== usage

  sql = "SELECT * FROM emp WHERE job = /*ctx[:job]*/'CLERK' AND deptno = /*ctx[:deptno]*/20"
  template = TwoWaySQL::Template.parse(sql)

  merged = template.merge(:job => "FOO", :deptno => 30)
  merged.sql                #=> "SELECT * FROM emp WHERE job = ? AND deptno = ?"
  merged.bound_variables    #=> ["FOO", 30]



==== IN clause

To bind multiple values in an IN clause, you can also use bind variable comment as well.

  IN /*argument name*/(...)

TwoWaySQL may use bind variable as follows. In this case, ctx[:names] is automatically replaced with placeholders associated with number of values in the data.

  IN /*ctx[:names]*/('aaa', 'bbb')

acceptable argument for IN clause is an array-like object. Say, Object that respond_to 'to_ary'.

===== usage

  sql = "SELECT * FROM emp WHERE deptno IN /*ctx[:deptnoList]*/(10, 20) ORDER BY ename"
  template = TwoWaySQL::Template.parse(sql)

  merged = template.merge(:deptnoList => [30,40,50])
  merged.sql                #=> "SELECT * FROM emp WHERE deptno IN (?, ?, ?) ORDER BY ename"
  merged.bound_variables    #=> [30,40,50]

  merged2 = template.merge(:deptnoList => [80])
  merged2.sql                #=> "SELECT * FROM emp WHERE deptno IN (?) ORDER BY ename"
  merged2.bound_variables    #=> [80]


==== LIKE

If you want to use "LIKE", you may write bind variables:

  ename LIKE /*ctx[:ename]*/'foo'

Unfortunately, there is no special support for "LIKE". So, to use a wildcard character, add wildcard directy to the data. For example, to specify to include "COT", add wildcard character in the value as follows:

  :ename => "%COT%"



==== Embedded variable comment

You can use Embedded variable comment to embed value directly (say without quoting or escaping) into the SQL as a string. Literal to the right of the Embedded variable comment will be replaced with value. Embedded variable comment has the following syntax:

  /*$variable name*/Literal

===== CAUTION:
As you noticed. Embedded variable comment has risk for SQL Injection. Please note, like any other 'eval' usage of TwoWaySQL, Embedded variable comment evals the data in safe level 4. Therefore, dangerous actions in Ruby world (ex. system call, valiable assignments, tainted strings) are never executed. However, this is not enough. Valid ruby string is still dangerous string as SQL fragments. Do NOT use user input or any other strings outside your code. If you use Embedded variable comment, you should carefully check the data and its origin.

===== usage

  sql = "SELECT * FROM emp ORDER BY /*$ctx[:order_by]*/ename /*$ctx[:order]*/ASC"
  template = TwoWaySQL::Template.parse(sql)

  merged = template.merge(:order_by => 'id, :order => 'DESC')
  merged.sql                #=> "SELECT * FROM emp ORDER BY id DESC"
  merged.bound_variables    #=> []



=== IF comment

To change SQL during execution based on a condition, use IF comments. IF comment has the following syntax:

  /*IF condition*/ .../*END*/

An example of IF comment is as follows:

  /*IF ctx[:foo]*/foo = /*ctx[:foo]*/'abc'/*END*/

When the condition returns a truthy value, TwoWaySQL treats statements in "/*IF*/" and "/*END*/" as active. In the above example, "foo = /*ctx[:foo]*/'abc'" will be output only when 'eval(ctx[:foo])' retuens an truthy value.


==== usage

  sql = "SELECT * FROM emp/*IF ctx[:job] */ WHERE job = /*ctx[:job]*/'CLERK'/*END*/"
  template = TwoWaySQL::Template.parse(sql)


  # active case
  merged = template.merge(:job => 'MANAGER')
  merged.sql                 #=> 'SELECT * FROM emp WHERE job = ?'
  merged.bound_variables     #=> ['MANAGER']

  # inactive case
  ctx = {}
  merged2 = template.merge(ctx)
  merged2.sql                #=> 'SELECT * FROM emp'
  merged2.bound_variables    #=> []



=== ELSE comment

You can use ELSE comment to activate statements when condition is false. Sn example of IF comment with ELSE is as follows.

  /*IF ctx[:foo]*/foo = /*ctx[:foo]*/'abc'
    -- ELSE foo IS NULL
  /*END*/

In this case, when the eval(ctx[:foo]) returns an falsy value, string "foo IS NULL" will be active.


==== ELSE comment sample

  sql = "SELECT * FROM emp WHERE /*IF ctx[:job]*/job = /*ctx[:job]*/'CLERK'-- ELSE job IS NULL/*END*/"
  template = TwoWaySQL::Template.parse(sql)

  # active case
  merged = template.merge(:job => 'MANAGER')
  merged.sql                 #=> 'SELECT * FROM emp WHERE job = ?'
  merged.bound_variables     #=> ['MANAGER']

  # inactive case
  ctx = {}
  merged2 = template.merge(ctx)
  merged2.sql                #=> 'SELECT * FROM emp WHERE job IS NULL'
  merged2.bound_variables    #=> []




=== BEGIN comment

BEGIN comment is used to not output WHERE clause when all IF comment in a WHERE clause, which does not include an ELSE, is false. BEGIN comment should used with IF comment.

BEGIN comment syntax is as follows:

  /*BEGIN*/WHERE clause/*END*/

So, BEGIN comment example is as follows:

  /*BEGIN*/WHERE
    /*IF ctx[:job]*/job = /*ctx[:job]*/'CLERK'/*END*/
    /*IF ctx[:deptno]*/AND deptno = /*ctx[:deptno]*/20/*END*/
  /*END*/

In the above example, 
* when job and deptno are nil, WHERE clause will not be outputted.
* When ctx[:job] == nil and ctx[:deptno] != nil, then sql will "WHERE depno = ?".
* When ctx[:job] != nil and ctx[:deptno] == nil, then sql will "WHERE job = ?".
* When ctx[:job] != nil and ctx[:deptno] != nil, then sql will "WHERE job = ? AND depno = ?".


==== usage

  sql = "SELECT * FROM emp/*BEGIN*/ WHERE /*IF ctx[:job]*/job = /*ctx[:job]*/'CLERK'/*END*//*IF ctx[:deptno]*/ AND deptno = /*ctx[:deptno]*/20/*END*//*END*/"
  template = TwoWaySQL::Template.parse(sql)

  # when data is empty (no param exists)
  ctx = {}
  merged = template.merge(ctx)
  merged.sql                 #=> 'SELECT * FROM emp'
  merged.bound_variables     #=> []

  # when :job param exists
  merged2 = template.merge(:job => 'MANAGER')
  merged2.sql                #=> 'SELECT * FROM emp WHERE job = ?'
  merged2.bound_variables    #=> ['MANAGER']

  # when :job and :deptno param exists
  ctx3 = {}
  ctx3[:job] = "MANAGER"
  ctx3[:deptno] = 20
  merged3 = template.merge(ctx3)
  merged3.sql                #=> 'SELECT * FROM emp WHERE job = ? AND deptno = ?'
  merged3.bound_variables    #=> ['MANAGER',20]



=== Parse Options

TwoWaySQL::Template.parse takes parse options as optional second argument. Acceptable parse options are kind_of Hash with available keys. Unknown options are just ignored.

* available parse option keys
  * :preserve_space (default is true)
  * :preserve_comment (default is false)
  * :debug (internal use only)


==== :preserve_space (default is true)

Default is true. When true, parser preserves original whitespaces. When false, parser translates consecutive whitespaces to single whitespace. This flag is useful for log space saving.


    sql = <<-EOS
  SELECT
    *
  FROM
    emp
  WHERE
    job    =   /*ctx[:job]*/'CLERK'
    AND   deptno   =   /*ctx[:deptno]*/10
  EOS
    template = TwoWaySQL::Template.parse(sql, :preserve_space => false)
    
    merged = template.merge(:job => 'MANAGER', :deptno => 30)
    
    merged.sql                 #=> "SELECT * FROM emp WHERE job = ? AND deptno = ? "
    merged.bound_variables     #=> ["MANAGER", 30]


==== :preserve_comment (default is false)

Default is false. When true, parser preserves original actual comments. When false, parser skips actual comment, therefore parsed SQL does not contain actual comments.

    sql = <<-EOS
  SELECT
    *
  FROM
    emp
    /* 
       This is a
       multiline comment
    */
  WHERE
    job    =   /*ctx[:job]*/'CLERK'
    AND   deptno   =   /*ctx[:deptno]*/10
  EOS
    template = TwoWaySQL::Template.parse(sql, :preserve_comment => true)
    
    merged = template.merge(:job => 'MANAGER', :deptno => 30)
    
    expected = <<-EOS
  SELECT
    *
  FROM
    emp
    /* 
       This is a
       multiline comment
    */
  WHERE
    job    =   ?
    AND   deptno   =   ?
  EOS
    merged.sql == expected     #=> true
    merged.bound_variables     #=> ["MANAGER", 30]





== REQUIREMENTS:

* racc/parser (basically bundled with ruby)


== INSTALL:

* (sudo) gem install twowaysql


== AUTHOR:

* Takuto Wada(takuto.wada at gmail dot com)


== CONTRIBUTORS:

* Keiji Muraishi(http://github.com/kjim)
* Keiji Yoshimi(http://github.com/walf443)


== LICENSE:

Copyright 2004-2008 the Seasar Foundation and the Others.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, 
either express or implied. See the License for the specific language
governing permissions and limitations under the License.


History.txt
== release_0_6 / 2015-04-03
* support ruby 2.1 and 2.2 (Thanks tim-saki)

== release_0_5 / 2012-01-20
* use Bundler to build this gem
* fix old specs / features

== release_0_4_1 / 2008-11-02
* apply latest newgem structure

== release_0_4 / 2008-09-29
* line number support on parse error
* integrate coverage report into website
* integrate ditz to 'push-button release'
* Change ambiguous class name TwoWaySQL::Result to TwoWaySQL::MergeResult

== release_0_3 / 2008-09-10
* better whitespace handling
* Whitespace compaction mode
* remove preserve_eol flag
* preserve actual comment content
* write docs for space compaction mode
* change line-based scanning to whole string based scanning

== 0.2.1 / 2008-09-07
* bugfix: case insensitive match for SQL reserved words
* make Ruby1.9 ready
* integrate ditz html into site
* change idiom 'substitution' to 'interpolation'
* Handle Embedded variable comment at Scanner/Grammar level

== 0.2.0 / 2008-09-05
* Initial release
* RubyForge settings
* rdoc for published classes and methods


spec/spec_helper.rb
begin
  require 'spec'
rescue LoadError
  require 'rubygems'
  gem 'rspec'
  require 'spec'
end

$:.unshift(File.dirname(__FILE__) + '/../lib')
require 'twowaysql'



lib/twowaysql/parser.rb
#
# DO NOT MODIFY!!!!
# This file is automatically generated by Racc 1.4.7
# from Racc grammer file "".
#

require 'racc/parser.rb'
module TwoWaySQL
  class Parser < Racc::Parser

module_eval(<<'...end parser.y/module_eval...', 'parser.y', 134)

require 'strscan'

def initialize(opts={})
  opts = {
    :debug => false,
    :preserve_space => true,
    :preserve_comment => false
  }.merge(opts)
  @yydebug = opts[:debug]
  @preserve_space = opts[:preserve_space]
  @preserve_comment = opts[:preserve_comment]
  @num_questions = 0
end


PAREN_EXAMPLE                = '\([^\)]+\)'
BEGIN_BIND_VARIABLE          = '(\/|\#)\*([^\*]+)\*\1'
BIND_VARIABLE_PATTERN        = /\A#{BEGIN_BIND_VARIABLE}\s*/
PAREN_BIND_VARIABLE_PATTERN  = /\A#{BEGIN_BIND_VARIABLE}\s*#{PAREN_EXAMPLE}/
EMBED_VARIABLE_PATTERN       = /\A(\/|\#)\*\$([^\*]+)\*\1\s*/

CONDITIONAL_PATTERN     = /\A(\/|\#)\*(IF)\s+([^\*]+)\s*\*\1/
BEGIN_END_PATTERN       = /\A(\/|\#)\*(BEGIN|END)\s*\*\1/
STRING_LITERAL_PATTERN  = /\A(\'(?:[^\']+|\'\')*\')/   ## quoted string
SPLIT_TOKEN_PATTERN     = /\A(\S+?)(?=\s*(?:(?:\/|\#)\*|-{2,}|\(|\)|\,))/  ## stop on delimiters --,/*,#*,',',(,)
LITERAL_PATTERN         = /\A([^;\s]+)/
SPACES_PATTERN          = /\A(\s+)/
QUESTION_PATTERN        = /\A\?/
COMMA_PATTERN           = /\A\,/
LPAREN_PATTERN          = /\A\(/
RPAREN_PATTERN          = /\A\)/
ACTUAL_COMMENT_PATTERN          = /\A(\/|\#)\*(\s{1,}(?:.*?))\*\1/m  ## start with spaces
SEMICOLON_AT_INPUT_END_PATTERN  = /\A\;\s*\Z/
UNMATCHED_COMMENT_START_PATTERN = /\A(?:(?:\/|\#)\*)/

#TODO: remove trailing spaces for S2Dao compatibility, but this spec sometimes causes SQL bugs...
ELSE_PATTERN            = /\A\-{2,}\s*ELSE\s*/
AND_PATTERN             = /\A(\ *AND)\b/i
OR_PATTERN              = /\A(\ *OR)\b/i


def parse( io )
  @q = []
  io.each_line(nil) do |whole|
    @s = StringScanner.new(whole)
  end
  scan_str

  # @q.push [ false, nil ]
  @q.push [ false, [@s.pos, nil] ]
    
  ## call racc's private parse method
  do_parse
end


## called by racc
def next_token
  @q.shift
end


def scan_str
  until @s.eos? do
    case
    when @s.scan(AND_PATTERN)
      @q.push [ :AND, [@s.pos, @s[1]] ]
    when @s.scan(OR_PATTERN)
      @q.push [ :OR, [@s.pos, @s[1]] ]
    when @s.scan(SPACES_PATTERN)
      @q.push [ :SPACES, [@s.pos, @s[1]] ]
    when @s.scan(QUESTION_PATTERN)
      @q.push [ :QUESTION, [@s.pos, nil] ]
    when @s.scan(COMMA_PATTERN)
      @q.push [ :COMMA, [@s.pos, ','] ]
    when @s.scan(LPAREN_PATTERN)
      @q.push [ :LPAREN, [@s.pos, '('] ]
    when @s.scan(RPAREN_PATTERN)
      @q.push [ :RPAREN, [@s.pos, ')'] ]
    when @s.scan(ELSE_PATTERN)
      @q.push [ :ELSE, [@s.pos, nil] ]
    when @s.scan(ACTUAL_COMMENT_PATTERN)
      @q.push [ :ACTUAL_COMMENT, [@s.pos, @s[1], @s[2]] ] if @preserve_comment
    when @s.scan(BEGIN_END_PATTERN)
      @q.push [ @s[2].intern, [@s.pos, nil] ]
    when @s.scan(CONDITIONAL_PATTERN)
      @q.push [ @s[2].intern, [@s.pos, @s[3]] ]
    when @s.scan(EMBED_VARIABLE_PATTERN)
      @q.push [ :EMBED_VARIABLE, [@s.pos, @s[2]] ]
    when @s.scan(PAREN_BIND_VARIABLE_PATTERN)
      @q.push [ :PAREN_BIND_VARIABLE, [@s.pos, @s[2]] ]
    when @s.scan(BIND_VARIABLE_PATTERN)
      @q.push [ :BIND_VARIABLE, [@s.pos, @s[2]] ]
    when @s.scan(STRING_LITERAL_PATTERN)
      @q.push [ :STRING_LITERAL, [@s.pos, @s[1]] ]
    when @s.scan(SPLIT_TOKEN_PATTERN)
      @q.push [ :IDENT, [@s.pos, @s[1]] ]
    when @s.scan(UNMATCHED_COMMENT_START_PATTERN)   ## unmatched comment start, '/*','#*'
      raise Racc::ParseError, "unmatched comment. line:[#{line_no(@s.pos)}], str:[#{@s.rest}]"
    when @s.scan(LITERAL_PATTERN)   ## other string token
      @q.push [ :IDENT, [@s.pos, @s[1]] ]
    when @s.scan(SEMICOLON_AT_INPUT_END_PATTERN)
      #drop semicolon at input end
    else
      raise Racc::ParseError, "syntax error at or near line:[#{line_no(@s.pos)}], str:[#{@s.rest}]"
    end
  end
end


## override racc's default on_error method
def on_error(t, v, vstack)
  ## cursor in value-stack is an array of two items,
  ##   that have position value as 0th item. like [731, "ctx[:limit] "]
  cursor = vstack.find do |tokens|
    tokens.size == 2 and tokens[0].kind_of?(Fixnum)
  end
  pos = cursor[0]
  line = line_no(pos)
  rest = @s.string[pos .. -1]
  raise Racc::ParseError, "syntax error at or near line:[#{line}], str:[#{rest}]"
end


def line_no(pos)
  lines = 0
  scanned = @s.string[0..(pos)]
  scanned.each_line { lines += 1 }
  lines
end
...end parser.y/module_eval...
##### State transition tables begin ###

racc_action_table = [
     8,    36,     9,    37,    12,    13,    10,    11,    14,    15,
    16,    17,    18,    19,    22,    23,    24,     8,    45,     9,
    46,    12,    13,    10,    11,    14,    15,    16,    17,    18,
    19,    22,    23,    24,     8,    40,     9,    25,    12,    13,
    10,    11,    14,    15,    16,    17,    18,    19,    22,    23,
    24,     8,    38,     9,     3,    12,    13,    10,    11,    14,
    15,    16,    17,    18,    19,    22,    23,    24,     8,   nil,
     9,   nil,    12,    13,    10,    11,    14,    15,    16,    17,
    18,    19,    22,    23,    24,    35,    33,    34,    31,    32,
    31,    32,    44,    43 ]

racc_action_check = [
    42,    24,    42,    24,    42,    42,    42,    42,    42,    42,
    42,    42,    42,    42,    42,    42,    42,     2,    37,     2,
    39,     2,     2,     2,     2,     2,     2,     2,     2,     2,
     2,     2,     2,     2,    41,    28,    41,     3,    41,    41,
    41,    41,    41,    41,    41,    41,    41,    41,    41,    41,
    41,    26,    26,    26,     1,    26,    26,    26,    26,    26,
    26,    26,    26,    26,    26,    26,    26,    26,    27,   nil,
    27,   nil,    27,    27,    27,    27,    27,    27,    27,    27,
    27,    27,    27,    27,    27,    22,    22,    22,     9,     9,
    40,    40,    34,    34 ]

racc_action_pointer = [
   nil,    54,    15,    37,   nil,   nil,   nil,   nil,   nil,    82,
   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,
   nil,   nil,    77,   nil,    -7,   nil,    49,    66,    30,   nil,
   nil,   nil,   nil,   nil,    84,   nil,   nil,    10,   nil,    17,
    84,    32,    -2,   nil,   nil,   nil,   nil,   nil ]

racc_action_default = [
    -2,   -35,    -1,   -35,    -3,    -4,    -5,    -6,    -2,    -2,
   -16,   -17,   -18,   -19,   -20,   -21,   -22,   -23,   -24,   -25,
   -26,   -27,   -35,   -32,   -35,    48,   -35,   -13,   -10,   -11,
   -12,    -2,    -2,   -28,   -35,   -30,   -33,   -35,    -7,   -35,
    -2,   -14,   -15,   -29,   -31,   -34,    -8,    -9 ]

racc_goto_table = [
     2,     1,    28,    39,   nil,   nil,   nil,   nil,    26,   nil,
   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,
   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,
   nil,    41,    42,    47 ]

racc_goto_check = [
     2,     1,     7,     8,   nil,   nil,   nil,   nil,     2,   nil,
   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,
   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,
   nil,     2,     2,     7 ]

racc_goto_pointer = [
   nil,     1,     0,   nil,   nil,   nil,   nil,    -7,   -25,   nil,
   nil,   nil,   nil ]

racc_goto_default = [
   nil,   nil,    27,     4,     5,     6,     7,   nil,   nil,    29,
    30,    20,    21 ]

racc_reduce_table = [
  0, 0, :racc_error,
  1, 20, :_reduce_1,
  0, 21, :_reduce_2,
  2, 21, :_reduce_3,
  1, 22, :_reduce_none,
  1, 22, :_reduce_none,
  1, 22, :_reduce_none,
  3, 25, :_reduce_7,
  4, 24, :_reduce_8,
  2, 27, :_reduce_9,
  0, 27, :_reduce_10,
  1, 26, :_reduce_none,
  1, 26, :_reduce_none,
  1, 26, :_reduce_none,
  2, 28, :_reduce_14,
  2, 29, :_reduce_15,
  1, 23, :_reduce_16,
  1, 23, :_reduce_17,
  1, 23, :_reduce_18,
  1, 23, :_reduce_19,
  1, 23, :_reduce_20,
  1, 23, :_reduce_21,
  1, 23, :_reduce_22,
  1, 23, :_reduce_23,
  1, 23, :_reduce_24,
  1, 23, :_reduce_25,
  1, 23, :_reduce_none,
  1, 23, :_reduce_none,
  2, 30, :_reduce_28,
  3, 30, :_reduce_29,
  2, 30, :_reduce_30,
  3, 30, :_reduce_31,
  1, 30, :_reduce_32,
  2, 31, :_reduce_33,
  3, 31, :_reduce_34 ]

racc_reduce_n = 35

racc_shift_n = 48

racc_token_table = {
  false => 0,
  :error => 1,
  :BEGIN => 2,
  :END => 3,
  :IF => 4,
  :ELSE => 5,
  :AND => 6,
  :OR => 7,
  :IDENT => 8,
  :STRING_LITERAL => 9,
  :SPACES => 10,
  :COMMA => 11,
  :LPAREN => 12,
  :RPAREN => 13,
  :QUESTION => 14,
  :ACTUAL_COMMENT => 15,
  :BIND_VARIABLE => 16,
  :PAREN_BIND_VARIABLE => 17,
  :EMBED_VARIABLE => 18 }

racc_nt_base = 19

racc_use_result_var = true

Racc_arg = [
  racc_action_table,
  racc_action_check,
  racc_action_default,
  racc_action_pointer,
  racc_goto_table,
  racc_goto_check,
  racc_goto_default,
  racc_goto_pointer,
  racc_nt_base,
  racc_reduce_table,
  racc_token_table,
  racc_shift_n,
  racc_reduce_n,
  racc_use_result_var ]

Racc_token_to_s_table = [
  "$end",
  "error",
  "BEGIN",
  "END",
  "IF",
  "ELSE",
  "AND",
  "OR",
  "IDENT",
  "STRING_LITERAL",
  "SPACES",
  "COMMA",
  "LPAREN",
  "RPAREN",
  "QUESTION",
  "ACTUAL_COMMENT",
  "BIND_VARIABLE",
  "PAREN_BIND_VARIABLE",
  "EMBED_VARIABLE",
  "$start",
  "sql",
  "stmt_list",
  "stmt",
  "primary",
  "if_stmt",
  "begin_stmt",
  "sub_stmt",
  "else_stmt",
  "and_stmt",
  "or_stmt",
  "bind_var",
  "embed_var" ]

Racc_debug_parser = false

##### State transition tables end #####

# reduce 0 omitted

module_eval(<<'.,.,', 'parser.y', 6)
  def _reduce_1(val, _values, result)
                      result = RootNode.new( val[0] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 11)
  def _reduce_2(val, _values, result)
                      result = []
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 15)
  def _reduce_3(val, _values, result)
                      result.push val[1]
                
    result
  end
.,.,

# reduce 4 omitted

# reduce 5 omitted

# reduce 6 omitted

module_eval(<<'.,.,', 'parser.y', 24)
  def _reduce_7(val, _values, result)
                      result = BeginNode.new( val[1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 29)
  def _reduce_8(val, _values, result)
                      result = IfNode.new( val[0][1], val[1], val[2] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 34)
  def _reduce_9(val, _values, result)
                      result = val[1]
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 38)
  def _reduce_10(val, _values, result)
                      result = nil
                
    result
  end
.,.,

# reduce 11 omitted

# reduce 12 omitted

# reduce 13 omitted

module_eval(<<'.,.,', 'parser.y', 47)
  def _reduce_14(val, _values, result)
                      result = SubStatementNode.new( val[0][1], val[1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 52)
  def _reduce_15(val, _values, result)
                      result = SubStatementNode.new( val[0][1], val[1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 57)
  def _reduce_16(val, _values, result)
                      result = LiteralNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 61)
  def _reduce_17(val, _values, result)
                      result = LiteralNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 65)
  def _reduce_18(val, _values, result)
                      result = LiteralNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 69)
  def _reduce_19(val, _values, result)
                      result = LiteralNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 73)
  def _reduce_20(val, _values, result)
                      result = WhiteSpaceNode.new( val[0][1], @preserve_space )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 77)
  def _reduce_21(val, _values, result)
                      result = LiteralNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 81)
  def _reduce_22(val, _values, result)
                      result = LiteralNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 85)
  def _reduce_23(val, _values, result)
                      result = LiteralNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 89)
  def _reduce_24(val, _values, result)
                      @num_questions += 1
                  result = QuestionNode.new( @num_questions )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 94)
  def _reduce_25(val, _values, result)
                      result = ActualCommentNode.new( val[0][1] , val[0][2] )
                
    result
  end
.,.,

# reduce 26 omitted

# reduce 27 omitted

module_eval(<<'.,.,', 'parser.y', 101)
  def _reduce_28(val, _values, result)
                      result = BindVariableNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 105)
  def _reduce_29(val, _values, result)
                      result = BindVariableNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 109)
  def _reduce_30(val, _values, result)
                      result = BindVariableNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 113)
  def _reduce_31(val, _values, result)
                      result = BindVariableNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 117)
  def _reduce_32(val, _values, result)
                      result = ParenBindVariableNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 122)
  def _reduce_33(val, _values, result)
                      result = EmbedVariableNode.new( val[0][1] )
                
    result
  end
.,.,

module_eval(<<'.,.,', 'parser.y', 126)
  def _reduce_34(val, _values, result)
                      result = EmbedVariableNode.new( val[0][1] )
                
    result
  end
.,.,

def _reduce_none(val, _values, result)
  val[0]
end

  end   # class Parser
  end   # module TwoWaySQL
